### ISSUE TITLE
Support for Etherpad plugins?

### ISSUE THREAD

**Issue by ar-jan** [kentonv/etherpad-lite#9]
Hi, I'm trying Etherpad on Sandstorm. It's working very nicely.

Are you planning on also supporting Etherpad plugins?

**Comment by kentonv** [comment]
Hi Arjan,

Interesting. Somehow I had not been aware that Etherpad _had_ plugins, which is obviously a pretty big oversight.

Sandstorm places each Etherpad document in an isolated container (which protects you against security bugs in Etherpad, of which there have been several). Unfortunately this means that simply enabling Etherpad's admin interface to let you install plugins won't really work -- you'd have to install the plugin again for every document, which is silly.

We'd like to add a feature to Sandstorm which allows "add-ons" to an app to be packaged as separate `.spk` files. We could then publish all the Etherpad plugins as such "add-ons", and they'd apply across all your documents. But, we don't have this yet.

For the time being, are there any particular plugins you'd particularly like to see on Sandstorm? We could simply install some set of plugins directly into the base Sandstorm package.

**Comment by ar-jan** [comment]
Hi Kenton, that sounds great.

I'd like to have available the styles dropdown (normal, headings, code), as seen on beta.etherpad.org, but I'm not sure if this is the headings plugin or something else.

That demo also has font size, font family and font color options. While I feel these could be useful, the web developer in me would rather see reusable (custom) styles and document-wide settings. I don't think that's currently possible. Do you think those font options would be good, or would it be maybe too much UI clutter to provide by default?

**Comment by kentonv** [comment]
I think we should definitely add the font and style options. I'll do that as soon as I get a chance to build a new package, hopefully soon. (Wish I'd realize these existed earlier.)

Custom styles would be neat but if I understand correctly (which I might not!) they currently can't be configured through a UI; you have to edit config files. That obviously doesn't work well with Sandstorm. Even if there were a UI, it would be kind of annoying to have to configure them for each document. Some day we could have the ability to create a set of styles in one grain and then apply them to documents through powerbox interactions... that's a ways out, though.

**Comment by kentonv** [comment]
I've just pushed a new Etherpad package which includes lots of formatting options, plus sidebar authorship highlighting (which I think looks a lot nicer) and comments (with the ability to grant comment permission independent of read and write). Let me know what you think!

**Comment by ar-jan** [comment]
Excellent! Indeed the sidebar authorship highlighting is much better. I'm also happy with the other additions. The separate comment permission looks very useful.

I noticed that if you share a View-only link, you still have all the built-in options like bold, italic. These are applied to the document, though not saved.

I looks like with the sidebar authorship, the Pad setting "Authorship colors" doesn't do anything (so maybe should be removed?). (Also, in the previous version I noticed that this Pad setting was not persistent for a document).

Thanks for the updates!

**Comment by kentonv** [comment]
> I noticed that if you share a View-only link, you still have all the built-in options like bold, italic. These are applied to the document, though not saved.

Seems to be an upstream bug.

> I looks like with the sidebar authorship, the Pad setting "Authorship colors" doesn't do anything

Hmm, I can't reproduce this. The sidebar authorship still uses coloring (weirdly, only while the edit area is focused), and updating authorship colors does in fact change these colors.

> Pad setting was not persistent for a document

Probably because cookies will be lost between sessions. Maybe we should fix this by saving cookies automatically in Sandstorm. Hmm...

**Comment by ar-jan** [comment]
> Hmm, I can't reproduce this. The sidebar authorship still uses coloring (weirdly, only while the edit area is focused), and updating authorship colors does in fact change these colors.

Yeah, changing the colors works for me too, I meant that before the sidebar-authorship was added, the checkbox under Pad settings "Authorship colors" enabled/disabled authorship altogether. Now that this information is in the sidebar that setting doesn't have any effect, as far as I can see.

**Comment by kentonv** [comment]
Ohh, I see what you mean. Yeah, that seems like a bug in the sidebar authorship plugin...

**Comment by ar-jan** [comment]
OK, looks like everything's upstream then, except maybe "Pad settings" persistence.

**Comment by ovizii** [comment]
I can confirm, the settings do not get saved, opened a separate issue for it too.
